src/humanloop/decorators/prompt.py

```diff
@@ -0,0 +1,149 @@
+from humanloop.types.model_providers import ModelProviders
+from humanloop.types.prompt_kernel_request_stop import PromptKernelRequestStop
+from humanloop.types.prompt_kernel_request_template import PromptKernelRequestTemplate
+from humanloop.types.response_format import ResponseFormat
+
+logger = logging.getLogger("humanloop.sdk")
+
+
+def prompt(
+    opentelemetry_tracer: Tracer,
+    path: Optional[str] = None,
+    # TODO: Template can be a list of objects?
+    model: Optional[str] = None,
+    attributes: Optional[dict[str, Any]] = None,
+    endpoint: Optional[ModelEndpoints] = None,
+    template: Optional[PromptKernelRequestTemplate] = None,
+    provider: Optional[ModelProviders] = None,
+    max_tokens: Optional[int] = None,
+    stop: Optional[PromptKernelRequestStop] = None,
+    temperature: Optional[float] = None,
+    top_p: Optional[float] = None,
+    presence_penalty: Optional[float] = None,
+    frequency_penalty: Optional[float] = None,
+    other: Optional[dict[str, Optional[Any]]] = None,
+    seed: Optional[int] = None,
+    response_format: Optional[ResponseFormat] = None,
+    tools: Optional[Sequence["ToolFunctionParams"]] = None,
+):
+    def decorator(func: Callable):
+        prompt_kernel = {}
+
+        if temperature is not None:
+            if not 0 <= temperature < 1:
+                raise ValueError(f"{func.__name__}: Temperature parameter must be between 0 and 1")
+            prompt_kernel["temperature"] = temperature
+
+        if top_p is not None:
+            if not 0 <= top_p <= 1:
+                raise ValueError(f"{func.__name__}: Top-p parameter must be between 0 and 1")
+            prompt_kernel["top_p"] = top_p
+
+        if presence_penalty is not None:
+            if not -2 <= presence_penalty <= 2:
+                raise ValueError(f"{func.__name__}: Presence penalty parameter must be between -2 and 2")
+            prompt_kernel["presence_penalty"] = presence_penalty
+
+        if frequency_penalty is not None:
+            if not -2 <= frequency_penalty <= 2:
+                raise ValueError(f"{func.__name__}: Frequency penalty parameter must be between -2 and 2")
+            prompt_kernel["frequency_penalty"] = frequency_penalty
+
+        for attr_name, attr_value in {
+            "model": model,
+            "endpoint": endpoint,
+            "template": template,
+            "provider": provider,
+            "max_tokens": max_tokens,
+            "stop": stop,
+            "other": other,
+            "seed": seed,
+            "response_format": response_format,
+            # {} -> None
+            "attributes": attributes or None,
+            "tools": tools,
+        }.items():
+            prompt_kernel[attr_name] = attr_value  # type: ignore
+
+        @wraps(func)
+        def wrapper(*args: Sequence[Any], **kwargs: Mapping[str, Any]) -> Any:
+            span: Span
+            with opentelemetry_tracer.start_as_current_span(str(uuid.uuid4())) as span:
```

Comment:
is uuid4 a convention for span IDs?

we should refactor this to be a single method too. (instead of importing and calling uuid everywhere)

Reply:
addressed